rfd/0098-registered-openssh-nodes.md

```diff
@@ -0,0 +1,90 @@
+
+#### Detecting OpenSSH nodes
+
+##### Option 1 - add information to SSH host certs
+
+When SSH host certificates are created for OpenSSH nodes, we set add an extension `x-teleport-role` that specifies what Teleport component the certificate is for. We could add another role type for OpenSSH nodes, so that when a Proxy connects to an OpenSSH node and reviews it's certificate, it could see from the `x-teleport-role` extension that the node is a registered OpenSSH node and preform the RBAC check appropriately.
+
+This would require adding a new `api/types.SystemRole` constant which would be backwards compatible.
+
+Pros:
+
+- Potentially easier implementation, setting and checking an SSH certificate extension is very easy.
+- Certificate fields couldn't be spoofed without access to a cluster's CAs, so the Proxy/Auth servers could be reasonably sure what type of node is what so RBAC checks wouldn't be skipped.
+
+Cons:
+
+- Old Teleport components may reject certificates with an unknown value for the `x-teleport-role` extension. `(api/types.SystemRoles).Check` will return with an error if it does not recognize the `api/type.SystemRole` constant.
+
+##### Option 2 - add a `node` resource sub-kind
+
+A new sub-kind to the `node` resource would be introduced. The name isn't important, but for now let's call this sub-kind `openssh`. When OpenSSH nodes are registered, this `openssh` sub-kind would have to be present in the resource. Then when a Proxy connects to a registered OpenSSH node it could lookup the sub-kind of the node, preforming the RBAC check appropriately. The absence of a `node` resource sub-kind would imply that the node is a Teleport agent node, making this change backwards compatible.
+
+Pros:
+
+- Node certificate generation doesn't need to be changed at all, so this change won't affect any files on remote nodes.
+
+Cons:
+
+- `lib/srv.Authhandlers` might need to be given access to the backend database in order to lookup the sub-kind of `node` resources so it can decide if the RBAC check can be done on the node or not.
+
+##### Option choice: 1
+
+I believe Option 1 is the better option, as it should be easier to implement and won't involve certificate validation code needing to access the backend database. Option 2 is certainly viable though, I honestly had a hard time deciding which option was the better one.
```

Comment:
Hm, option 2 seems like it fits better in our existing architecture, but maybe that's because I don't fully understand option 1.

Reply:
I didn't fully understand our SSH certificate authentication flow, after talking with @jakule we came to the conclusion that Option 1 is not viable. I originally assumed that when connecting to and authenticating with an OpenSSH node, the Host certificate was the only certificate that would be used for authentication so some info could be added to it. But both Host and User certificates are used for authentication, and node-specific info shouldn't be added to a user's certificate.